app/controllers/hideout_controller.rb

```diff
@@ -0,0 +1,94 @@
+class HideoutController < ApplicationController
+  def create
+    # Needs owner_email and name as params.
+    # name must be lesser than 20 characters and greater than 3 characters
+    # owner must not already be in a hideout
+    # return the hideout as an object
+    owner_email = params[:owner_email]
+    name = params[:name]
+
+    if !User.exists?(email: owner_email)
+      render status: :not_found, body: "User does not exist"
+    elsif name.length > 20
+      render status: :precondition_failed, body: "Hideout name greater than 20 characters"
+    elsif name.length < 3
+      render status: :precondition_failed, body: "Hideout name lesser than 3 characters"
+    elsif !User.find_by(email: owner_email).hideout_id.nil?
+      render status: :precondition_failed, body: "Owner already in hideout"
+    else
+      hideout_id = Hideout.create(owner_email, name)
+      render status: :created, :json => {name: name, owner_email: owner_email, hideout_id: hideout_id}
+    end
+  end
+
+  def destroy
+    hideout_id = params[:hideout_id]
+    issued_by_email = params[:user_email]
+
+    if !Hideout.exists?(id: hideout_id)
+      render status: :not_found, body: "Hideout does not exist"
+    elsif !User.exists?(email: issued_by_email)
+      render status: :not_found, body: "User does not exist"
+    elsif Hideout.find_by(id: hideout_id).owner_id != User.find_by(email: issued_by_email).id
+    render status: :forbidden, body: "User is not the Hideout owner"
+    else 
+      Hideout.destroy(hideout_id)
+      render status: :ok
+    end
+  end
+
+  def rename
+    hideout_id = params[:hideout_id]
+    issued_by_email = params[:user_email]
+    new_name = params[:new_name]
+
+    if !Hideout.exists?(id: hideout_id)
+      render status: :not_found, body: "Hideout does not exist"
+    elsif !User.exists?(email: issued_by_email)
+      render status: :not_found, body: "User does not exist"
+    elsif Hideout.find_by(id: hideout_id).owner_id != User.find_by(email: issued_by_email).id
+    render status: :forbidden, body: "User is not the Hideout owner"
```

Comment:
Ok I notice now that we're repeating these code checks for user / hideout existence. Here's what we can do. For now, remove the checks for the user does not exist. We will set up a middleware in the next sprint to check for user authorization using Firebase ID tokens that guarantees user existence in our database for any request before any of these functions are triggered.

As for this particular route collection, in the next we'll make a Hideout middleware that abstracts the hideout does not exist / user is owner checks into two separate middleware (one for hideout does not exist, one for user is owner). For now, just leave comments in each function specifying which middleware is required.